### ISSUE TITLE
Is it possible to make our search handle spaces, special characters and mistakes more lenient?

### ISSUE THREAD

**Issue by mietcls** [ugent-library/biblio-backoffice#1045]
# UX crash

Is it possible to make our search handle spaces, special characters and mistakes more lenient?
This is very important to search for doubles. Example:

Example 1: When searching for "health care" or "healthcare" the search returns different results.
Example 2: When searching for "D-4" or "D4" the search returns different results.
Example 3: When searching for ") :" or "):" the search returns different results.

# Images

Example 1
![Image](https://user-images.githubusercontent.com/1572885/228264510-c35f9add-10d4-422a-87a7-a888583be542.png)

Example 2
![Image](https://user-images.githubusercontent.com/1572885/228264630-f960b4e4-0098-4e5c-aa88-ad37a7ec5d24.png)

Example 3
![Image](https://user-images.githubusercontent.com/1572885/228264670-56538da2-8ce9-4622-8653-1dfe63c9e9d2.png)

# Context

Original message
> (1) Als ik de zin "Why Jecker's capabilities approach to age-based rationing is incapable of containing health care costs" in het zoekveld ingeef (met health care af elkaar), krijg ik niet de dubbel te zien wanneer ik de zin "Why Jecker's capabilities approach to age-based rationing is incapable of containing healthcare costs" "healthcare" aan elkaar vast.
> (2) Hier hetzelfde verhaal: D-4 versus D4.
> (3) laatste voorbeeld, nu met "spatie of geen spatie dubbelpunt".

**Comment by nicolasfranck** [comment]
That is one of the things that search indexes (like ElasticSearch or Solr) are not good at out of the box.

Most of the times people try to fix this using using synonym lists, but we do not know all cases (and neither do the users), and multi word synonyms have their own problems.

Search indexes are term oriented. They do not know that `D-4` and `D4` are the same, or should.

Spaces have meaning. How else could one extract search terms out of a full text string, and perform text analysis on it?

Databases on the other hand try to find such values by using all kinds of wildcard queries, but then fail at text analysis (i.e. match `text` when I type `texts` as it has the same base word).

**Comment by mietcls** [comment]
@nicolasfranck Let me rephrase: how will librarians be able to find approximate doubles easily?

**Comment by netsensei** [comment]
@mietcls Remember Pat's talk earlier this week? It's that kind of problem. @nicolasfranck is correct about the limits of search indices and database systems. 

Relevancy search in Solr or Elastic works by matching a search term against an index of text fragments and gauging the frequency of matches (TD-IDF algorithm). More matches means that a record has a higher relevancy and should get a higher search ranking.

"finding approximate doubles" is actually a form of clustering items in a set based on a statistical similarity factor. It's part of a higher-order problem class: gauging the **"similarity"** between items. Other sub-problems include: recommendations, anomaly detection, diversity measurement and classification.

So, gauging direct similarity between records is not the same as frequency based search using a set of search terms.

Typically, the idea is to quantify the data into a mathematical vector (a set of numbers e.g. `[4, 10, 9]`) and then rely on algorithms, statistics and math to calculate the distance between those vectors (e.g. Lehvenstein distance). Vectors that are closer to each other have a higher degree of similarity, then those who don't.

How does this work in the real world?

One approach is to use the `pgvector` (https://github.com/pgvector/pgvector/) plugin for PostgreSQL. Which allows you to write SQL queries that returns the similarity between rows in a SQL table with vector-based data. The big question then becomes: How do you convert text - titles, abstracts, author names,... - into vectors which accurately represent the the semantic complexity of the data?  In the AI/ML age, LLM's are leveraged to generate those vectors e.g. OpenAI offers an API that allows you to convert data into vectors / embeddings (https://platform.openai.com/docs/guides/embeddings/what-are-embeddings) At that point, you need to read the fine print that comes with those LLM's and how they interpret the data you feed them.

The caveat being that even then, in a set of several 100k records, you'll still have to deal with precision / recall to gauge the accuracy of your similarity querying.

Either way, it's a problem that doesn't have a clear-cut answer.

**Comment by nics** [comment]
To make this more actionable: we could make a periodic job that finds close matches and stores candidate duplicates as a relation or as a field inside the record. When a reviewer views a record she can see a list of possible duplicates.
Search would be limited to finding records that have possible duplicates and finding the possible duplicates for a given id.

I have scripts lying around that found levenshtein title matches in LibreCat, we could use that as a starting point.

**Comment by mietcls** [comment]
Okay thank you! I will start a doubles epic – to create prevention and antidote features.

**Comment by nicolasfranck** [comment]
@nics @netsensei agree with your remarks about detecting duplicates, but I think the scope of this ticket was merely about lenient behaviour of matching.

**Comment by mietcls** [comment]
Keeping track of this in #1099 until we have a better insight in both exact search and duplicate detection.
